### PR TITLE
Comments section for package reports.

### PR DESCRIPTION
**Pull request by LoisSotoLopez**

### CODE REVIEW

File: lib/hexpm/repository/comments.ex [hunk 1]
```diff
@@ -0,0 +1,18 @@
+defmodule Hexpm.Repository.Comments do
```
Comment (ericmj) [inline lib/hexpm/repository/comments.ex]: I think this would fit better under the `PackageReports` module.

File: lib/hexpm/repository/comment.ex [hunk 2]
```diff
@@ -0,0 +1,40 @@
+defmodule Hexpm.Repository.Comment do
```
Comment (ericmj) [inline lib/hexpm/repository/comment.ex]:
```suggestion
defmodule Hexpm.Repository.PackageReportComment do
```

File: lib/hexpm_web/controllers/package_report_controller.ex [hunk 3]
```diff
@@ -10,6 +10,23 @@ defmodule HexpmWeb.PackageReportController do
...
+    comments =
+      %{
+        "report" => report,
+        "author" => author,
+        "text" => params["text"]
+      }
+      |> Comments.new()
+
+    conn
+    |> redirect(to: Routes.package_report_path(HexpmWeb.Endpoint, :show, report.id))
```
Comment (ericmj) [inline lib/hexpm_web/controllers/package_report_controller.ex]:
```suggestion
    redirect(conn, to: Routes.package_report_path(HexpmWeb.Endpoint, :show, report.id))
```

We usually avoid single pipes. Same for `Comments.new` above.

File: lib/hexpm_web/controllers/package_report_controller.ex [hunk 4]
```diff
@@ -73,18 +90,31 @@ defmodule HexpmWeb.PackageReportController do
...
       |> redirect(to: Routes.package_path(HexpmWeb.Endpoint, :index))
     else
+      has_comments = Comments.count(report.id) != 0
+
+      comments =
+        if has_comments do
+          Comments.all_for_report(report.id)
+        else
+          []
+        end
```
Comment (ericmj) [inline lib/hexpm_web/controllers/package_report_controller.ex]: This is not an optimization. It's faster to just call `Comments.all_for_report(report.id)`.

File: lib/hexpm/repository/comment.ex [hunk 5]
```diff
@@ -0,0 +1,40 @@
+defmodule Hexpm.Repository.PackageReportComment do
+  use Hexpm.Schema
+  import Ecto.Query, only: [from: 2]
+
+  schema "comments" do
```
Comment (ericmj) [inline lib/hexpm/repository/comment.ex]: The table name should match the module name. The migrations needs to change also.

```suggestion
  schema "package_report_comments" do
```

File: lib/hexpm_web/controllers/package_report_controller.ex [hunk 6]
```diff
@@ -10,6 +10,22 @@ defmodule HexpmWeb.PackageReportController do
...
+    report = PackageReports.get(report_id)
+    author = conn.assigns.current_user
+
+    comments =
+      %{
+        "report" => report,
+        "author" => author,
+        "text" => params["text"]
+      }
+      |> Comments.new()
```
Comment (ericmj) [inline lib/hexpm_web/controllers/package_report_controller.ex]: Remove the pipe here.

File: lib/hexpm_web/controllers/package_report_controller.ex [hunk 7]
```diff
@@ -73,18 +89,31 @@ defmodule HexpmWeb.PackageReportController do
...
       |> redirect(to: Routes.package_path(HexpmWeb.Endpoint, :index))
     else
+      has_comments = Comments.count(report.id) != 0
+
+      comments =
+        if has_comments do
+          Comments.all_for_report(report.id)
+        else
+          []
+        end
```
Comment (ericmj) [inline lib/hexpm_web/controllers/package_report_controller.ex]: Looks like this hasn't been fixed?

File: lib/hexpm_web/templates/package_report/_comment.html.eex [hunk 8]
```diff
@@ -0,0 +1,5 @@
+
```
Comment (ericmj) [inline lib/hexpm_web/templates/package_report/_comment.html.eex]: Remove this line.

```suggestion
```

File: lib/hexpm_web/controllers/package_report_controller.ex [hunk 9]
```diff
@@ -10,6 +10,21 @@ defmodule HexpmWeb.PackageReportController do
   @report_bad_version_msg "No release matchs given requirement"
   @report_not_accessible "Requested package report not accessible"
 
+  def new_comment(conn, params) do
+    report_id = params["id"]
+    report = PackageReports.get(report_id)
+    author = conn.assigns.current_user
+
+
```
Comment (ericmj) [inline lib/hexpm_web/controllers/package_report_controller.ex]:
```suggestion
```

Run formatter :)

File: lib/hexpm_web/controllers/package_report_controller.ex [hunk 10]
```diff
@@ -73,18 +88,26 @@ defmodule HexpmWeb.PackageReportController do
...
     user = conn.assigns.current_user
+    for_owner = Owners.get(report.package, user) != nil
 
     if report == nil do
       conn
       |> put_flash(:error, @report_not_accessible)
       |> put_status(400)
       |> redirect(to: Routes.package_path(HexpmWeb.Endpoint, :index))
     else
+      has_comments = PackageReports.count_comments(report.id) != 0
```
Comment (ericmj) [inline lib/hexpm_web/controllers/package_report_controller.ex]: Do not pass this down to the template, instead you can do `@comments != []` in the template.

**Review by ericmj — APPROVED** [review]

**Comment by ericmj** [comment]
@LoisSotoLopez looks like you didn't push any changes since the last review

EDIT: I also just noticed that the tests are missing for this.

File: lib/hexpm_web/templates/package_report/show.html.eex [hunk 11]
```diff
@@ -2,14 +2,19 @@
...
 <%= if @for_moderator do %>
-    <%= render "_moderator_pannel.html", report: @report, author: @report.author%>
+  <%= render "_moderator_panel.html", report: @report, author: @report.author%>
 <% end%>
 
+<%= if @for_moderator or @for_owner do %>
+  <%= render "_comments_section.html", report: @report, author: @report.author, has_comments: @has_comments, comments: @comments%>
+<% end %>
+
+
```
Comment (ericmj) [inline lib/hexpm_web/templates/package_report/show.html.eex]: Remove empty lines.

**Comment by LoisSotoLopez** [comment]
I've added the needed access control  on the PackageReportController and the comments section template logic. Package owners should be able to see comments sections on reports about their packages once the report has been accepted, but should they be able to comment too?

**Comment by ericmj** [comment]
> Package owners should be able to see comments sections on reports about their packages once the report has been accepted, but should they be able to comment too?

Package owners should the report itself when it has been accepted, before that it is only visible to the moderators, so the comments should be included automatically when they can see the report. Package owners should be able to comment as well once the report has been accepted

**Comment by ericmj** [comment]
Can you add the tests for this functionality as well?

EDIT: Let us know if you need any help with that.

**Pull request merged**